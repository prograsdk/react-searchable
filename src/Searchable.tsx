import lodashDebounce from 'lodash.debounce';
import {ChangeEvent, Component, ReactNode} from 'react';

/**
 * A predicate function used for filtering items.
 *
 * @example
 * ```typescript
 *
 * interface User {
 *   name: string;
 *   email: string;
 * };
 *
 * const predicate: IFilteringPredicate<User> = (user, query) =>
 *   user.name.includes(query) || user.email.includes(query);
 * ```
 *
 * @typeparam T - The type of items that will be filtered using the predicate.
 * @param item - The current item in the filtering process.
 * @param query - A search query used for filtering.
 */
export interface IFilteringPredicate<T> {
  (item: T, query: string): boolean;
}

export interface IRenderProp<T> {
  (
    props: {
      items: T[];
      query: string;
      handleChange(event: ChangeEvent<HTMLInputElement>): void;
    },
  ): ReactNode;
}

interface IBaseProps<T> {
  /**
   * The duration for debouncing the filtering function.
   */
  debounce: number | boolean;
  /**
   * An initial search query. Will affect initial state.
   */
  initialQuery: string;
  /**
   * The array of items to filter.
   */
  items: T[];
  /**
   * The predicate used for filtering items.
   */
  predicate: IFilteringPredicate<T>;
}

interface IPropsWithChildren<T> extends IBaseProps<T> {
  children: IRenderProp<T>;
}

interface IPropsWithRender<T> extends IBaseProps<T> {
  render: IRenderProp<T>;
}

export type IProps<T> = IPropsWithChildren<T> | IPropsWithRender<T>;

interface IState<T> {
  /**
   * An array of filtered items based on [[IProps.items]].
   */
  items: T[];
  /**
   * Current query used for filtering.
   */
  query: string;
}

/**
 * @typeparam T - The type of items to search.
 */
export default class Searchable<T> extends Component<IProps<T>, IState<T>> {
  public static defaultProps = {
    debounce: 300,
    initialQuery: '',
  };

  /**
   * Filters an array of items based on a search query and a filtering predicate.
   *
   * @typeparam T - The type of items.
   * @param items - The array of items to be filtered.
   * @param query - A search string to be passed to the predicate.
   * @param predicate - A filtering predicate based on an item and the search query.
   * @returns A filtered array of items.
   */
  public static filter<T>(
    items: T[],
    query: string,
    predicate: IFilteringPredicate<T>,
  ) {
    return items.filter((item: T) => predicate(item, query));
  }

  constructor(props: IProps<T>) {
    super(props);

    const {initialQuery: query, predicate, items, debounce} = props;

    this.state = {
      items:
        query !== '' ? Searchable.filter<T>(items, query, predicate) : [],
      query,
    };

    if (debounce) {
      const duration =
        typeof debounce === 'number'
          ? debounce
          : Searchable.defaultProps.debounce;
      this.filterAndSetState = lodashDebounce(this.filterAndSetState, duration);
    }
  }

  /**
   * Change event handler for updating query in state.
   *
   * @param event - An instance of ChangeEvent
   */
  public handleChange = (event: ChangeEvent<HTMLInputElement>): void =>
    this.setState({query: event.target.value})

  /**
   * Call filtering method if [[IState.query]] has changed.
   */
  public componentDidUpdate(prevProps: IProps<T>, prevState: IState<T>): void {
    const {query} = this.state;

    if (query !== prevState.query) {
      const {items} = this.props;

      this.filterAndSetState(items, query);
    }
  }

  public render(): ReactNode {
    const {items, query} = this.state;

    const renderParams = {items, query, handleChange: this.handleChange};

    if (isPropsWithRender(this.props)) {
      return this.props.render(renderParams);
    } else {
      return this.props.children(renderParams);
    }
  }

  /**
   * Filters an array of items and sets state. This method might be debounced in [[constructor]].
   *
   * @param items - The array of items to filter.
   * @param query - The search query to base the filtering on.
   */
  public filterAndSetState(items: T[], query: string): void {
    const {predicate} = this.props;

    this.setState({
      items:
        query !== '' ? Searchable.filter<T>(items, query, predicate) : [],
    });
  }
}

function isPropsWithRender<T>(props: IProps<T>): props is IPropsWithRender<T> {
  return (props as IPropsWithRender<T>).render !== undefined;
}
