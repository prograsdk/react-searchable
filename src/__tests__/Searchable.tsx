import {mount, shallow, ShallowWrapper} from 'enzyme';
import debounce from 'lodash.debounce';
import React, {ReactNode} from 'react';
import Searchable, {IProps, IRenderProp} from '../Searchable';

jest.mock('lodash.debounce', (): any => jest.fn((fn: () => any) => fn));

interface IUser {
  name: string;
}

interface ISetup<T> {
  wrapper: ShallowWrapper;
  props: IProps<T>;
}

function setup(overrides?: any): ISetup<IUser> {
  const props: IProps<IUser> = {
    items: [
      {
        name: 'Stewart Harrison',
      },
      {
        name: 'Jake Bullock',
      },
    ],
    predicate: (item: IUser, query: string): boolean =>
      item.name.includes(query),
    ...overrides,
  };

  const wrapper: ShallowWrapper = shallow<Searchable<IUser>, {}, {}>(
    <Searchable<IUser> {...props}>{() => null}</Searchable>,
  );

  return {
    props,
    wrapper,
  };
}

describe('<Searchable />', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('rendering', () => {
    let render: IRenderProp<IUser>;

    beforeEach(() => {
      render = () => <p>searchable</p>;
    });

    it('renders without crashing', () => {
      const {wrapper} = setup();

      expect(wrapper).toMatchSnapshot();
    });

    it('renders children if passed', () => {
      const {props} = setup();

      const wrapper = mount(
        <Searchable<IUser> {...props}>{render}</Searchable>,
      );

      expect(wrapper.find('p').exists()).toBe(true);
    });

    it('renders render prop if passed', () => {
      const {props} = setup();

      const wrapper = mount(<Searchable<IUser> {...props} render={render} />);

      expect(wrapper.find('p').exists()).toBe(true);
    });
  });

  describe('debouncing', () => {
    it('debounces filtering function with default on true', () => {
      const {wrapper} = setup({debounce: true});
      const instance = wrapper.instance() as Searchable<IUser>;

      const filter = instance.filterAndSetState;

      expect(debounce).toHaveBeenCalledWith(
        filter,
        Searchable.defaultProps.debounce,
      );
    });

    it('debounces filtering function with pased duration', () => {
      const {wrapper} = setup({debounce: 500});
      const instance = wrapper.instance() as Searchable<IUser>;

      const filter = instance.filterAndSetState;

      expect(debounce).toHaveBeenCalledWith(filter, 500);
    });

    it('does not debounce the function on false', () => {
      const {wrapper} = setup({debounce: false});

      expect(debounce).not.toHaveBeenCalled();
    });
  });

  describe('filtering', () => {
    it('returns empty array if query is empty string', () => {
      const {
        wrapper,
        props: {items},
      } = setup({ initialQuery: 'corncob' });

      wrapper.setState({query: ''});

      expect(wrapper.state('items')).toEqual([]);
    });

    it('filters items based on predicate', () => {
      const {wrapper} = setup();

      wrapper.setState({query: 'Jake'});

      expect(wrapper.state('items')).toEqual([{name: 'Jake Bullock'}]);
    });

    it('filters initial items based on initialQuery', () => {
      const {wrapper} = setup({initialQuery: 'Jake'});

      expect(wrapper.state('items')).toEqual([{name: 'Jake Bullock'}]);
    });

    it('returns empty array if no matches', () => {
      const {wrapper} = setup();

      wrapper.setState({query: 'corncob'});

      expect(wrapper.state('items')).toEqual([]);
    });

    it('does not call Searchable.filter if query is empty', () => {
      const {wrapper} = setup();

      // Mock Searchable.filter
      const filter = jest.fn();
      Searchable.filter = filter;

      wrapper.setState({query: ''});

      expect(filter).not.toHaveBeenCalled();
    });
  });

  describe('handleChange', () => {
    it('sets state', () => {
      const {wrapper} = setup();
      const instance = wrapper.instance() as Searchable<IUser>;

      // tslint:disable:no-object-literal-type-assertion
      const event = {
        target: {value: 'Jake'},
      } as React.ChangeEvent<HTMLInputElement>;

      instance.handleChange(event);

      expect(wrapper.state('query')).toEqual('Jake');
    });
  });
});
