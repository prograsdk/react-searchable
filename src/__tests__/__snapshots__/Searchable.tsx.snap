// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Searchable /> rendering renders without crashing 1`] = `ShallowWrapper {}`;
